pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status api.ConditionStatus
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	Reason string
+	// Human-readable message indicating details about last transition.
+	Message string
+}
+
+// FlowSchemaConditionType is a valid value for FlowSchemaStatusCondition.Type
+type FlowSchemaConditionType string
+
+// FlowSchemaConditionStatus is a valid value for FlowSchemaStatusCondition.Status
+type FlowSchemaConditionStatus string
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfiguration represents the configuration of a priority level.
+type PriorityLevelConfiguration struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec PriorityLevelConfigurationSpec
+	// Current status of a request-priority.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status PriorityLevelConfigurationStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+	// Items is a list of request-priorities.
+	Items []PriorityLevelConfiguration
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// GlobalDefault specifies whether this priority level should be considered as the default priority for requests
+	// that do not have any priority level. Only one PriorityClass can be marked as `globalDefault`.
+	// +optional
+	GlobalDefault bool
+	// Exempt defines whether the priority level is exempted or not.  There should be at most one exempt priority level.
+	// Being exempt means that requests of that priority are not subject to concurrency limits (and thus are never queued)
+	// and do not detract from the concurrency available for non-exempt requests. In a more sophisticated system, the
+	// exempt priority level would be the highest priority level. The field is default to false and only those system
+	// preset priority level can be exempt.
+	// +optional
+	Exempt bool
+	// AssuredConcurrencyShares is a positive number for a non-exempt priority level, representing the weight by which
```

Comment:
Is it fair to say we only expect ~5 of these per cluster?

With low numbers like that, is there a benefit to presenting shares rather than just a percent?

At one point I suggested accepting either one, with whatever fraction is not consumed by the priorities requesting absolute percent divided up among those requesting shares.

I don't feel super strongly but I do think folks will be more comfortable specifying some things (e.g. system critical) in percent?

Reply:
I would prefer to start simple.  I suggest shares rather than percent because percents are fragile --- add or remove one priority level and you have to rejigger the percentages of the others.

Reply:
That's fine.

Reply:
percents will look indeed better in the model's json/yaml output, comparing to raw numbers. and rejiggering won't be that costy if there were only ~5?

> Imagine that cluster startup includes some process creating PriorityLevelConfiguration objects. 

not an issue for starting up, the initializing can be done in a post-start-hook, which is blocking/synchronous before the server starts working.

Reply:
and is there any usable work of percentage types in the codebase so far?

Reply:
It's not possible to change these atomically, and changing percents one by one may do strange things to the system. Let's stick with shares for the moment.

Reply:
alternatively, we can put the percentage into the status, so that users can see the change if they adjust the shares

Reply:
Just remember that multiple objects can not be updated at once, so the percentages in status might not always add up to 100.